<h2 align='center' style='color:purple'>Problem-Set2 Part 1</h2>

**For cancer dataset in sklearn library, we are going to find out best model and best hyper parameters using GridSearchCV and RandomisedSearchCV**

In [1]:
from sklearn import svm, datasets
cancer = datasets.load_breast_cancer()

In [2]:
cancer.keys()

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])

In [3]:
cancer

{'data': array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
         1.189e-01],
        [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
         8.902e-02],
        [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
         8.758e-02],
        ...,
        [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
         7.820e-02],
        [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
         1.240e-01],
        [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
         7.039e-02]], shape=(569, 30)),
 'target': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
        0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
        1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
        1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
        1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1,

#### Description

The dataset contains 569 samples with 30 features each. The features are real, positive values. The dataset is divided into two classes: malignant and benign.

In [4]:
import pandas as pd
df = pd.DataFrame(cancer.data,columns=cancer.feature_names)
df['cancer'] = cancer.target
df['cancer'] = df['cancer'].apply(lambda x: cancer.target_names[x])
df[47:50]

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
47        13.17         18.66           85.98      534.6          0.11580   
48        12.05         14.63           78.04      449.3          0.10310   
49        13.49         22.30           86.91      561.0          0.08752   

    mean compactness  mean concavity  mean concave points  mean symmetry  \
47           0.12310         0.12260              0.07340         0.2128   
48           0.09092         0.06592              0.02749         0.1675   
49           0.07698         0.04751              0.03384         0.1809   

    mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
47                 0.06777  ...          27.95           102.80       759.4   
48                 0.06043  ...          20.70            89.88       582.6   
49                 0.05718  ...          31.82            99.00       698.8   

    worst smoothness  worst compactness  worst concavity  \
47            0.1786             0.4166           0.5006   
48            0.1494             0.2156           0.3050   
49            0.1162             0.1711           0.2282   

    worst concave points  worst symmetry  worst fractal dimension     cancer  
47               0.20880          0.3900                  0.11790  malignant  
48               0.06548          0.2747                  0.08301     benign  
49               0.12820          0.2871                  0.06917     benign  

[3 rows x 31 columns]

<h3 style='color:blue'> train_test_split </h3>

In [5]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
X_train, X_test, y_train, y_test = train_test_split(cancer.data, cancer.target, test_size=0.2,random_state = 0)

In [6]:
model = svm.SVC(kernel='rbf',C=30,gamma='auto')
model.fit(X_train,y_train)
model.score(X_test, y_test)
y_pred = model.predict(X_test)

In [7]:
mse = mean_squared_error(y_test, y_pred)
print(f"Mean Squared Error: {mse}")
score = model.score(X_test, y_test)
print(f"Accuracy: {score}")

Mean Squared Error: 0.41228070175438597
Accuracy: 0.5877192982456141


In [8]:
from sklearn.ensemble import RandomForestClassifier
model2 = RandomForestClassifier(n_estimators=100, random_state=0)
model2.fit(X_train,y_train)
model2.score(X_test, y_test)
y_pred1 = model2.predict(X_test)

In [9]:
mse1 = mean_squared_error(y_test, y_pred1)
print(f"Mean Squared Error: {mse1}")
score1 = model2.score(X_test, y_test)
print(f"Accuracy: {score1}")

Mean Squared Error: 0.03508771929824561
Accuracy: 0.9649122807017544


<h3 style='color:blue'>Use GridSearchCV</h3>

In [10]:
from sklearn import svm
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import RandomizedSearchCV

model_params = {
    'svm': {
        'model': svm.SVC(gamma='auto'),
        'params' : {
            'C': [1,10,20],
            'kernel': ['rbf','linear']
        }
    },
    'random_forest': {
        'model': RandomForestClassifier(random_state = 0),
        'params' : {
            'n_estimators': [1,10,100]
        }
    }
}

In [11]:
scores = []

for model_name, mp in model_params.items():
    clf =  GridSearchCV(mp['model'], mp['params'], cv=5, return_train_score=False)
    clf.fit(cancer.data, cancer.target)
    scores.append({
        'model': model_name,
        'best_score': clf.best_score_,
        'best_params': clf.best_params_
    })

df = pd.DataFrame(scores,columns=['model','best_score','best_params'])
df

model  best_score                    best_params
0            svm    0.952569  {'C': 10, 'kernel': 'linear'}
1  random_forest    0.963111          {'n_estimators': 100}

<h3 style='color:blue'>Use RandomisedSearchCV</h3>

In [12]:
scores = []

for model_name, mp in model_params.items():
    rs = RandomizedSearchCV(mp['model'], mp['params'], cv=5, return_train_score=False,n_iter=2)
    rs.fit(cancer.data, cancer.target)
    scores.append({
        'model': model_name,
        'best_score': rs.best_score_,
        'best_params': rs.best_params_
    })

df = pd.DataFrame(scores,columns=['model','best_score','best_params'])
df

model  best_score                 best_params
0            svm    0.627418  {'kernel': 'rbf', 'C': 10}
1  random_forest    0.963111       {'n_estimators': 100}